modules/network/src/main/java/org/apache/ignite/network/scalecube/ScaleCubeTopologyService.java

```diff
@@ -54,37 +52,16 @@ void setLocalMember(Member member) {
     void onMembershipEvent(MembershipEvent event) {
         ClusterNode member = fromMember(event.member());
 
-        String memberName = member.name();
-
-        switch (event.type()) {
-            case ADDED:
-                members.put(memberName, member);
-
-                fireAppearedEvent(member);
-
-                break;
-
-            case LEAVING:
-                members.remove(memberName);
-
-                fireDisappearedEvent(member);
+        if (event.isAdded()) {
+            this.members.put(member.address(), member);
 
-                break;
-
-            case REMOVED:
-                // In case if member left non-gracefully, without sending LEAVING event.
-                if (members.remove(memberName) != null)
-                    fireDisappearedEvent(member);
-
-                break;
-
-            case UPDATED:
-                // No-op.
-                break;
-
-            default:
-                throw new IgniteInternalException("This event is not supported: event = " + event);
+            fireAppearedEvent(member);
+        }
+        else if (event.isRemoved()) {
+            this.members.compute(member.address(), // Ignore stale remove event.
+                (k, v) -> v.id().equals(member.id()) ? null : v);
 
+            fireDisappearedEvent(member);
```

Comment:
I guess this should not be called if the event is stale

Reply:
The event is "stale" only in the context of membership view. 
But, because each event now is bound to specific cluster id, it has to be fired or it will be lost.
For example, it's a correct event reordering:
join A(id=1)          join A(id=1)
fail A(id=1)   ---> join A(id=2)
join A(id=2)          fail(A(id=1)

Reply:
Thanks for explanation!